### DISCUSSION TITLE
Kiali-operator, kiali cr, grafana/prometheus auth

### DISCUSSION THREAD

**Discussion by edward738** [kiali/kiali#7793] [Q&A] [answered]
### Describe the bug
I'm using Kiali-operator
If Grafana or Prometheus needs credentials for login, Kiali  couldn't authenticate
Creds are valid, I can access  Grafana and Prometheus using these creds by curl

kiali cr config
<img width="511" alt="Grafana-auth" src="https://github.com/user-attachments/assets/e0d6bf4a-1baa-474c-a93e-e1e95d343fe4">
<img width="560" alt="prometheus-auth" src="https://github.com/user-attachments/assets/337db71e-a9e7-41c9-bbfd-49c3233794ca">

kiali logs
<img width="1062" alt="kiali-logs" src="https://github.com/user-attachments/assets/f676fd4b-913e-47e1-9fa3-45333ee0d877">
<img width="391" alt="kiali" src="https://github.com/user-attachments/assets/ff94917f-af66-42f0-9260-0a7a4309cc16">

### Expected Behavior
Kiali could access to the resources

### What are the steps to reproduce this bug?

1. Install kiali-operator by helm
2. Configure grafana or prometheus authentication
3. Set creds of prometheus or grafana into kiali cr 

### Environment

- **Kiali-operator  app version: 1.89.7
- **Kiali-operator  helm version: 1.89.7
- **Istio version: 1.23.1
- **Kubernetes version: v1.30.4+rke2r1

**Comment by jmazzitelli** [comment]
_(moving to Q&A discussion as this is most likely not a bug; we can move it to issues if it is confirmed a bug)_

All we can say at this point is you are getting a 401 which is bad credentials. The prometheus error about "bad_response" seems to indicate the url might not be pointing to the correct place.

**Reply by jmazzitelli** [reply]
Make sure the Kiali pod has access to those URLs. We've seen it before where the URLs provided either were not resolvable from within the Kiali pod or were not accessible to the Kiali pod.

**Reply by edward738** [reply]
Kiali pod has  access to those URLs and creds are valid. I can access it by curl and creds

**Reply by edward738** [reply]
If you specify in_cluster_url in this format
in_cluster_url: https://user:12345@hoost.name/
then everything works and access errors do not appear.
however it is not safe to use it this way
If you specify
username: user
password: 12345
in_cluster_url: https://host.name/
then we get the error 401 access denied
Perhaps the password field should be specified in some encrypted form, but this is not specified in the documentation.
Can you clarify whether any additional actions need to be performed on the password field or is this a bug in kiali

**Reply by jmazzitelli** [reply]
One thing that we've noticed very recently for something completely unrelated, but just struck me as I see your example above... would your username and/or password just happen to be all numerals? In your example above your password is `12345` - is that literally what it is?

If so, wrap it in quotes. Sometimes when unmarshalling YAML in the code, if all-numerals are not wrapped in quotes, the text is attempted to be cast as a number which causes problems.

Again, this was only relevant in something totally unrelated, but... perhaps a similar thing is happening here.

Make sure you wrap your values in quotes and try again.

**Reply by edward738** [reply]
it also doesn't work, I have made this 

My point is:
1. This block doesn't work
grafana:
  auth:
    password: "12345"
    username: "name"
  in_cluster_url: "http://host.name"
  
2. but this block works very well
grafana:
  in_cluster_url: "http://user:12345@host.name"

**Comment by jmazzitelli** [comment]
Note: Your grafana URLs in your screen shots were showing "http", not "https". That might be one issue.

But to answer your question, yes, we recommend people use secrets to more securely specify your passwords. See the Kiali documentation here: https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.external_services

**Reply by jmazzitelli** [reply]
Note the following can be stored in secrets using that method documented above:

- grafana-password
- grafana-token
- prometheus-password
- prometheus-token
- tracing-password
- tracing-token
- login-token-signing-key

**Comment by jmazzitelli** [comment, marked as answer]
:) The problem was staring me in the face and I didn't notice.

Your `auth.type` is incorrect. You have `none` which is telling Kiali there is no auth. You want `basic`.

Please see docs:

- https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.external_services.custom_dashboards.prometheus.auth.type
- https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.external_services.grafana.auth.type

**Reply by edward738** [reply]
Thanks a lot)
It helped
